File: 32. Longest Valid Parentheses/32.cpp
#include<stack>
#include<iostream>
using namespace std;

class Solution {
public:
    int longestValidParentheses(string s) {
        // last is the position of the last ')"
        int ans = 0, last = -1;
        // store all the indices of '('
        stack<int> stk;
        for(int i = 0; i < s.size(); i++) {
            if(s[i] == '(') {
                stk.push(i);
            } else {  
                // there is no matching '("
                if(stk.empty()) {
                    last = i;
                } else {
                    stk.pop();
                    if(stk.empty()) {
                        ans = max(ans,i-last);
                    } else {
                        ans = max(ans,i-stk.top());
                    }
                }
            }
        }
        return ans;
    }
};

int main()
{
    Solution *s = new Solution;
    cout<<s->longestValidParentheses("())()(()");
    return 0;
}

File: 32. Longest Valid Parentheses/32.py
class Solution:
    # @param {string} s
    # @return {integer}
    def longestValidParentheses(self, s):
        dp = [0] * (len(s)+1)           # dp[i]: longest length of s[:i]
        count = 0                       # count is the num of unmatched '('
        for i in range(1,len(dp)):
            if s[i-1] == '(':
                count += 1
            else:
                if count:
                    count -= 1
                    dp[i] = 2
                    if i - 2 >= 0 and s[i-2] == ')':
                        dp[i] += dp[i-1]
                    dp[i] += dp[i-dp[i]]
        print(dp)
        return max(dp)

Solution.longestValidParentheses(Solution,'(()())())')
